### PR TITLE
[FIX] [IMPROVE] Prevent creation of duplicated roles and new `roles.update` endpoint

### PR DESCRIPTION
**Pull request by lucassartor**
## Proposed changes (including videos or screenshots)

Currently, the action of updating a role is broken: because roles have their `_id` = `name`, when updating a role there's no way to validate if the user is trying to update or create a new role with a name that already exists -  which causes wrong behaviors, such as roles with the same name and not being able to update them.

To proper fix this, this PR looks to change the creation of roles. Now, roles have a unique  `_id` value and there's a endpoint to update roles: `/api/v1/roles.update`.

Doing so, it's possible to validate on both endpoints (`roles.create` and `roles.update`) to not allow roles with duplicated names.

**OBS:** The unique id changes only reflect new roles, the standard roles (such as admin and user) still have `_id` = `name`, but new roles now **can't** have the same name as them.

## Issue(s)

[ClickUp Task](https://app.clickup.com/t/512jnp)

## Further comments

This PR only fixes the behavior of the endpoints. The UI still uses the `authorization:saveRole` method to create/update roles, which is still broken and will probably be **deprecated**.

### CODE REVIEW

File: app/api/server/v1/roles.js [hunk 1]
```diff
@@ -44,12 +46,29 @@ API.v1.addRoute('roles.create', { authRequired: true }, {
...
+
+		if (Roles.findOneByIdOrName(roleData.name)) {
+			throw new Meteor.Error('error-duplicate-role-names-not-allowed', 'Role name already exists');
+		}
+
+		if (['Users', 'Subscriptions'].includes(roleData.scope) === false) {
+			roleData.scope = 'Users';
+		}
+
+		const roleId = Roles.createWithRandomId(roleData.name, roleData.scope, roleData.description, false, roleData.mandatory2fa);
```
Comment (pierre-lehnen-rc) [inline app/api/server/v1/roles.js:61]: `roleData.mandatory2fa` needs to be declared on the check's schema for it to be a valid param (first line of the method).

File: app/api/server/v1/roles.js [hunk 2]
```diff
@@ -104,3 +123,53 @@ API.v1.addRoute('roles.getUsersInRole', { authRequired: true }, {
...
+API.v1.addRoute('roles.update', { authRequired: true }, {
+	post() {
+		check(this.bodyParams, {
+			roleId: String,
+			name: Match.Maybe(String),
+			scope: Match.Maybe(String),
+			description: Match.Maybe(String),
+		});
+
+		const roleData = {
```
Comment (pierre-lehnen-rc) [inline app/api/server/v1/roles.js:136]: no option to update the `mandatory2fa` attribute?

File: app/models/server/models/Roles.js [hunk 3]
```diff
@@ -28,6 +28,32 @@ export class Roles extends Base {
 		});
 	}
 
+	updateById(_id, name, scope, description) {
```
Comment (pierre-lehnen-rc) [inline app/models/server/models/Roles.js:31]: no option to update the `mandatory2fa` attribute?

File: app/api/server/v1/roles.js [hunk 4]
```diff
@@ -104,3 +123,53 @@ API.v1.addRoute('roles.getUsersInRole', { authRequired: true }, {
...
+			scope: this.bodyParams.scope,
+			description: this.bodyParams.description,
+		};
+
+		const role = Roles.findOneByIdOrName(roleData.roleId);
+
+		if (!role) {
+			throw new Meteor.Error('error-invalid-roleId', 'This role does not exist');
+		}
+
```
Comment (pierre-lehnen-rc) [inline app/api/server/v1/roles.js:148]: the update should also throw an error if the user is trying to modify a protected role.

**Review by pierre-lehnen-rc — CHANGES_REQUESTED** [review]

File: app/api/server/v1/roles.js (hunk 2 above)
Comment (lucassartor) [reply app/api/server/v1/roles.js:136]: I thought the `mandatory2FA` should not be updated by the user via the API as the `roles.create` also doesn't have this parameter, my bad! I'll fix it, thanks for noticing.

**Comment by KevLehman** [comment]
What would prevent a user from creating a role with the same name as one of the default ones? 🤔

**Comment by lucassartor** [comment]
> What would prevent a user from creating a role with the same name as one of the default ones? 🤔

On `roles.update` this validation: https://github.com/RocketChat/Rocket.Chat/blob/6de6f36f0e67a998534a84911c65be6503ea51cc/app/api/server/v1/roles.js#L155#L160

And on `role.create` this one: https://github.com/RocketChat/Rocket.Chat/blob/6de6f36f0e67a998534a84911c65be6503ea51cc/app/api/server/v1/roles.js#L53#L55

**Review by pierre-lehnen-rc — CHANGES_REQUESTED** [review]
Can you fix the 'roles.create' endpoint to also accept the mandatory2fa param?

**Review by pierre-lehnen-rc — APPROVED** [review]

**Pull request merged**